storagemarket/impl/provider_environments.go

```diff
@@ -48,48 +85,58 @@ func (p *providerDealEnvironment) Ask() storagemarket.StorageAsk {
 	return *sask.Ask
 }
 
-func (p *providerDealEnvironment) DeleteStore(storeID multistore.StoreID) error {
-	return p.p.multiStore.Delete(storeID)
-}
-
-func (p *providerDealEnvironment) GeneratePieceCommitment(storeID *multistore.StoreID, payloadCid cid.Cid, selector ipld.Node, pieceSize abi.PaddedPieceSize) (cid.Cid, filestore.Path, error) {
-	proofType, err := p.p.spn.GetProofType(context.TODO(), p.p.actor, nil)
+// GeneratePieceCommitment generates the pieceCid for the CARv1 deal payload in
+// the CARv2 file that already exists at the given path.
+func (p *providerDealEnvironment) GeneratePieceCommitment(proposalCid cid.Cid, carPath string, dealSize abi.PaddedPieceSize) (c cid.Cid, path filestore.Path, finalErr error) {
+	rd, err := carv2.OpenReader(carPath)
 	if err != nil {
-		return cid.Undef, "", err
+		return cid.Undef, "", xerrors.Errorf("failed to get CARv2 reader, proposalCid=%s, carPath=%s: %w", proposalCid, carPath, err)
 	}
 
-	var pieceCid cid.Cid
-	var path filestore.Path
-	var psize abi.UnpaddedPieceSize
+	defer func() {
+		if err := rd.Close(); err != nil {
+			log.Errorf("failed to close CARv2 reader, carPath=%s, err=%s", carPath, err)
+
+			if finalErr == nil {
+				c = cid.Undef
+				path = ""
+				finalErr = xerrors.Errorf("failed to close CARv2 reader, proposalCid=%s, carPath=%s: %w",
+					proposalCid, carPath, err)
+				return
+			}
+		}
+	}()
 
-	if p.p.universalRetrievalEnabled {
-		pieceCid, psize, path, err = providerutils.GeneratePieceCommitmentWithMetadata(p.p.fs, p.p.pio.GeneratePieceCommitment, proofType, payloadCid, selector, storeID)
-	} else {
-		pieceCid, psize, err = p.p.pio.GeneratePieceCommitment(proofType, payloadCid, selector, storeID)
+	// dump the CARv1 payload of the CARv2 file to the Commp Writer and get back the CommP.
```

Comment:
If `rd.DataReader()` is an `io.Reader`,  the `io.Copy` can be avoided by just using padreader and the FFI directly. Not blocking for merge but this seems like and optimization worth pursuing.

All the complicated machinery from before is due to having to write the CAR file from the blockstore and then read it to compute CommP. Here we already have a reader.

Cribbed from source code inside writer --

``` golang
import (
	ffi "github.com/filecoin-project/filecoin-ffi"
	"github.com/filecoin-project/go-padreader"
)

r, sz := padreader.New(rd.DataReader(), rd.Header.DataSize)
pieceCid, err := ffiwrapper.GeneratePieceCIDFromFile(abi.RegisteredSealProof_StackedDrg32GiBV1, r, sz)
pieceSize := sz.Padded()
```

Reply:
@dirkmc can you follow-up on this?

Reply:
@raulk please take a look 👁️👁️

Reply:
Two things here, @hannahhoward:

1. Are you suggesting to stop using `go-commp-utils/writer.Writer` in favour of this more direct approach? I see that `writer.Writer` implements some parallelisation. Does the recipe you suggest also do that? Sorry but I'm not familiar with the motivation and the existence of these utilities, and unfortunately they are not documented either.
2. Are you sure you want a direct `filecoin-ffi` dependency from `go-fil-markets`? Right now it's not on the direct build graph.

Reply:
Confirmed offline with @hannahhoward that we don't want to stop using `writer.Writer`.